compiler/ras/ILValidator.hpp

```diff
@@ -0,0 +1,214 @@
+         {
+         int32_t index = _sideTable.indexOf(node);
+         TR_ASSERT(index >= _basis, "Cannot mark node n%dn before basis %d live", node->getGlobalIndex(), _basis);
+         _liveOffsets.set(index - _basis);
+         }
+
+      void remove(TR::Node *node)
+         {
+         int32_t index = _sideTable.indexOf(node);
+         if (index >= _basis)
+            _liveOffsets.reset(index - _basis);
+         }
+
+      bool isEmpty()
+         {
+         return _liveOffsets.isEmpty();
+         }
+
+      TR::Node *anyLiveNode()
+         {
+         Iterator iter(*this);
+         return iter.currentNode();
+         }
+
+      void startNewWindow()
+         {
+         TR_ASSERT(_liveOffsets.isEmpty(), "Cannot close LiveNodeWindow when there are still live nodes");
+         _basis = _sideTable.size();
+         }
+
+      public:
+
+      class Iterator
+         {
+         LiveNodeWindow      &_window;
+         TR_BitVectorIterator _bvi;
+         int32_t              _currentOffset; // -1 means iteration is past end
+
+         bool isPastEnd(){ return _currentOffset < 0; }
+
+         public:
+
+         Iterator(LiveNodeWindow &window):_window(window),_bvi(window._liveOffsets),_currentOffset(0xdead /* anything non-negative */)
+            {
+            stepForward();
+            }
+
+         TR::Node *currentNode()
+            {
+            if (isPastEnd())
+               return NULL;
+            else
+               return _window._sideTable.getAt(_window._basis + _currentOffset)._node;
+            }
+
+         void stepForward()
+            {
+            TR_ASSERT(!isPastEnd(), "Can't stepForward a LiveNodeWindow::Iterator that's already past end");
+            if (_bvi.hasMoreElements())
+               _currentOffset = _bvi.getNextElement();
+            else
+               _currentOffset = -1;
+            }
+
+         public: // operators
+
+         void operator ++() { stepForward(); }
+         };
+      };
+
+   NodeSideTable<NodeState>  _nodeStates;
+   LiveNodeWindow            _liveNodes;
+
+   bool isValidSoFar(){ return _isValidSoFar; }
+   void validityRule(Location &location, bool condition, const char *formatStr, ...);
+
+   protected: // Individual validation routines
+
+   // These return false to indicate a validity error was found.
+   // Note that these are given verb names ("validateXxx") instead of adjective
+   // names ("xxxIsValid") because generally they are stateful.
+
+   // Called on every node, before all other validation routines
+   //
+   virtual void validateNode(Location &location);
+
+   //
+   // Additional validation hooks
+   //
+
+   virtual void validateEndOfExtendedBlock(Location &location);
+
+   protected:
+
+   TR::Compilation *comp();
+
+   void updateNodeState(Location &newLocation);
+
+   void printDiagnostic(const char *formatStr, ...);
+   void vprintDiagnostic(const char *formatStr, va_list ap);
+   };
+
+class JavaILValidator: public ILValidator
```

Comment:
Originally, `JavaILValidator` was in `omr/compiler/infra/ILWalk.{hpp,cpp}`. It just got moved around as part of my refactoring. I've removed it from the OMR compiler though and it will now live as part of the Java project.